### PR TITLE
[nexus] Add virtual provisioning idempotency tests, prevent underflow

### PR DESCRIPTION
**Pull request by smklein**
Builds on https://github.com/oxidecomputer/omicron/pull/5081 and https://github.com/oxidecomputer/omicron/pull/5089 , but more out of convenience than necessity.

# Summary

This PR attempts to validate that, for the "virtual provisioning collection {insert, delete}" operations, they are idempotent. Currently, our usage of `max_instance_gen` only **partially** prevents updates during instance provisioning deletions:
- If `max_instance_gen` is smaller than the observed instance generation number...
- ... we avoid deleting the `virtual_provisioning_resource` record (which is great)
- ... but we still decrement the `virtual_provisioning_collection` values (which is really not great).

This basically means that we can "only cause the project/silo/fleet usage values to decrement arbitrarily, with no other changes". This has been, mechanically, the root cause of our observed underflows (e.g, https://github.com/oxidecomputer/omicron/issues/5525). 

# Details of this change

- All the changes in `nexus/db-queries/src/db/datastore/virtual_provisioning_collection.rs` are tests validating idempotency of these operations.
- All the changes in `nexus/db-queries/src/db/queries/virtual_provisioning_collection_update.rs` are actually changes to the query which change functionality. The objective of these changes is to preserve idempotency of the newly added tests, and to prevent undercounting of virtual provisioning resources. If these changes are reverted, the newly added tests start failing, showing a lack of coverage.

### CODE REVIEW

File: nexus/db-queries/src/db/queries/virtual_provisioning_collection_update.rs [hunk 1]
```diff
@@ -239,11 +239,52 @@ WITH
...
+                // resources if the instance generation was already advanced to or
+                // past G.
+                //
+                // If the relevant instance ID is not in the database, then some
+                // other operation must have ensured the instance was previously
+                // stopped (because that's the only way it could have been deleted),
+                // and that operation should have cleaned up the resources already,
+                // in which case there's nothing to do here.
+                query.sql("
+  do_update
```
Comment (smklein) [inline nexus/db-queries/src/db/queries/virtual_provisioning_collection_update.rs:262]: I basically tried to move all the validation of the state_generation into this boolean `do_update?` subquery, then re-use that check in multiple spots where we can perform updates below.

File: nexus/db-queries/tests/output/virtual_provisioning_collection_update_delete_instance.sql [hunk 2]
```diff
@@ -40,25 +40,17 @@ WITH
             1
         )
         = 1
+        AND EXISTS(
+            SELECT 1 FROM instance WHERE instance.id = $5 AND instance.state_generation < $6 LIMIT 1
+          )
```
Comment (smklein) [inline nexus/db-queries/tests/output/virtual_provisioning_collection_update_delete_instance.sql:45]: The problem here is that (before this PR) `do_update` could be "true", but the state generation could be below our "max", which would result in that buggy partial deletion state.

Now:
- If `do_update` is true, we really believe we should perform the insertion/deletion
- If that's the case, we can use it as the deciding factor to both delete the `virtual_provisioning_resource` **and** decrement the usage in the `virtual_provisioning_collection` table

**Review by gjcolombo — APPROVED** [review]
Thanks for doing this! LGTM as far as the use of the `do_update` clause goes--this is more or less the fix I had in mind for #5042, which issue I think this will resolve.

I know the conversion to raw SQL is part of a different PR, but I also looked at the entire expected query output in `virtual_provisioning_collection_update_delete_instance.sql` and the structure there also looks right to me (I didn't check the parameter ordering too closely though).

File: nexus/db-queries/src/db/datastore/virtual_provisioning_collection.rs [hunk 3]
```diff
@@ -325,3 +325,614 @@ impl DataStore {
...
+        datastore
+            .silo_update_quota(&opctx, &authz_silo, quotas_update)
+            .await
+            .unwrap();
+
+        // Actually provision the instance
+
+        let instance_id = Uuid::new_v4();
+        let cpus = 12;
+        let ram = ByteCount::try_from(1 << 30).unwrap();
```
Comment (hawkw) [inline nexus/db-queries/src/db/datastore/virtual_provisioning_collection.rs:464]: nit, take it or leave it: it looks like creating the instance record is factored out into a function, but the rest of this test setup code is repeated for all the tests. why not just have one `test_setup` (or something) that does all of this?

**Review by hawkw — APPROVED** [review]

File: nexus/db-queries/src/db/datastore/virtual_provisioning_collection.rs (hunk 3 above)
Comment (smklein) [reply nexus/db-queries/src/db/datastore/virtual_provisioning_collection.rs]: Good feedback, I'll incorporate this. I was being lazy!

**Pull request merged**